### PR TITLE
trying to fix PAM regarding push monitoring

### PR DESCRIPTION
**Pull request by kovacshuni**

### CODE REVIEW

File: notifications.go [hunk 1]
```diff
@@ -1,5 +1,7 @@
 package main
 
+const ChangeType = "http://www.ft.com/thing/ThingChangeType/"
```
Comment (Nesze) [inline notifications.go]: Should this be exported (I know, this is not a lib, but still) ?

File: handler.go [hunk 2]
```diff
@@ -12,9 +12,10 @@ const evPrefix = "data: "
 const errMsgPrefix = "Serving /notifications request: [%v]"
 
 type handler struct {
-	dispatcher         *eventDispatcher
-	notificationsCache queue
-	apiBaseUrl         string
+	dispatcher           *eventDispatcher
+	notificationsCache   queue
+	apiBaseUrl           string
+	notificationsBaseUrl string
```
Comment (Nesze) [inline handler.go]: Do we need this explicitly? If yes, could we come up with a better naming, like `internalBaseUrl` or something else?

If no, you can extract the `links` creation into a separate function in the `notifications` handler

Comment (kovacshuni) [inline handler.go]: Renamed.

File: notifications.go (hunk 1 above)
Comment (kovacshuni) [inline notifications.go]: lowercased :)

File: app.go [hunk 3]
```diff
@@ -97,7 +97,7 @@ func main() {
 		infoLogger.Printf("Config: [\n\tconsumerAddrs: [%v]\n\tconsumerGroupID: [%v]\n\ttopic: [%v]\n\tconsumerAutoCommitEnable: [%v]\n\tapiBaseURL: [%v]\n\tnotifications_capacity: [%v]\n]", *consumerAddrs, *consumerGroupID, *topic, *consumerAutoCommitEnable, *apiBaseURL, *nCap)
 
 		notificationsCache := newCircularBuffer(*nCap)
-		h := handler{dispatcher, notificationsCache, *apiBaseURL}
+		h := handler{dispatcher, notificationsCache, *apiBaseURL, *apiBaseURL + "/__notifications-push"}
```
Comment (Nesze) [inline app.go]: Just a suggestion, extracting this into a builder function, would simplify things, as you would not need to pass the last parameter here, nor in the tests.

File: queue.go [hunk 4]
```diff
@@ -2,33 +2,46 @@ package main
...
 type queue interface {
-	enqueue(i item)
-	dequeue() item
-	items() []item
+	enqueue(*notificationUPP)
+	dequeue() *notificationUPP
+	items() []*notificationUPP
 }
 
 type circularBuffer struct {
```
Comment (Nesze) [inline queue.go]: (technically, it's not a circularBuffer anymore)

File: queue.go [hunk 5]
```diff
@@ -2,33 +2,46 @@ package main
...
 
 func newCircularBuffer(capacity int) queue {
-	return &circularBuffer{&sync.Mutex{}, make([]item, 0, capacity)}
+	return &circularBuffer{&sync.Mutex{}, make([]*notificationUPP, 0, capacity), capacity}
 }
 
-func (cb *circularBuffer) enqueue(i item) {
+func (cb *circularBuffer) enqueue(n *notificationUPP) {
+	wasRemoved := true
+	for wasRemoved {
```
Comment (Nesze) [inline queue.go]: do you need to nested fors here?

**Comment by Nesze** [comment]
Please run `go fmt`, and some linters, `golint` or even `gometalinter`

File: queue.go [hunk 6]
```diff
@@ -2,33 +2,46 @@ package main
...
 }
 
-func (cb *circularBuffer) enqueue(i item) {
+func (cb *arrayUniqueue) enqueue(n *notificationUPP) {
+	wasRemoved := true
+	for wasRemoved {
+		wasRemoved = false
+		for i, e := range cb.buffer {
+			if (e.ID == n.ID && e.Type == n.Type) {
+				cb.buffer = append(cb.buffer[:i], cb.buffer[i+1:]...)
```
Comment (Nesze) [inline queue.go]: potential `slice bounds out of range` panic

https://play.golang.org/p/750O8iw3Go

File: app.go (hunk 3 above)
Comment (kovacshuni) [inline app.go]: done

File: queue.go (hunk 5 above)
Comment (kovacshuni) [inline queue.go]: discussed, remained.

File: queue.go (hunk 4 above)
Comment (kovacshuni) [inline queue.go]: renamed

**Comment by Nesze** [comment]
👍

**Pull request merged**